substrate/frame/staking/src/pallet/impls.rs

```diff
@@ -1843,14 +1875,44 @@ impl<T: Config> Pallet<T> {
 		Self::check_exposures()?;
 		Self::check_paged_exposures()?;
 		Self::check_ledgers()?;
+		Self::check_locks()?;
 		Self::check_count()
 	}
 
+	/// Invariants:
+	///
+	/// * Staking locked funds for every bonded stash should be the same as its ledger's total.
+	///
+	/// NOTE: some of these checks result in warnings only. Once
+	/// <https://github.com/paritytech/polkadot-sdk/issues/3245> is resolved, turn warns into check
+	/// failures.
+	fn check_locks() -> Result<(), TryRuntimeError> {
+		for (_, controller) in <Bonded<T>>::iter() {
+			if let Some(ledger) = <Ledger<T>>::get(&controller) {
+				let lock = <T::Currency as InspectLockableCurrency<T::AccountId>>::balance_locked(
+					crate::STAKING_ID,
+					&ledger.stash,
+				);
+
+				if lock != ledger.total {
+					// TODO before merging: use warn
```

Comment:
We don't catch warnings quite well yet, Error is actually safer, why switch to warn?

Reply:
The rationale is to throw warns until all the try state checks pass in Polkadot and Kusama to avoid raising alarms on a problem we already know about. And then revert to errs once the state is fixed https://github.com/paritytech/polkadot-sdk/issues/3728.

Reply:
However, I turned this check into an `ensure!`, so it will throw an error.